### PR TITLE
Bug 1232720 - Create a make dist command that begins with making a production ready xpi.

### PR DESCRIPTION
**Pull request by Standard8**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -49,6 +43,9 @@ distserver: remove_old_config dist
 
 BUILT := ./built
 ADD-ON := add-on
+BUILT_ADD_ON := ./built/$(ADD-ON)
```
Comment (Mardak) [inline Makefile]: Probably should reference `$(BUILT)` instead of `./built`

File: Makefile [hunk 2]
```diff
@@ -49,6 +43,9 @@ distserver: remove_old_config dist
...
+DIST := ./dist
+XPI_NAME := loop@mozilla.org.xpi
```
Comment (Mardak) [inline Makefile:48]: Could move  +XPI_FILE := built/$(XPI_NAME) up to here with the rest. Also $(BUILT)

File: Makefile [hunk 3]
```diff
@@ -171,6 +179,53 @@ $(BUILT)/$(ADD-ON)/chrome.manifest: $(ADD-ON)/jar.mn
...
+# - Don't copy prefs.js as we generate that in the dependencies.
+# - Copy all the relevant js, html, css etc files that we need.
+# - Copy the add-on "support files".
+# - Exclude all other files.
+# - Exclude all test files.
+# - Exclude and editor temp files.
+RSYNC_DIST := $(RSYNC) -r \
+  --exclude="**/prefs.js" \
+  --include="*.jsm" --include="*.js" \
+	--include="*.html" --include="*.css"  \
```
Comment (Mardak) [inline Makefile]: nit: got a mix of spaces and tabs here. and extra space before \

File: Makefile [hunk 4]
```diff
@@ -171,6 +179,53 @@ $(BUILT)/$(ADD-ON)/chrome.manifest: $(ADD-ON)/jar.mn
...
+add-on_dist: $(DIST)/add-on/chrome/content/preferences/prefs.js
+	mkdir -p $(DIST)/add-on
+	# Copy just the files we need.
+	$(RSYNC_DIST) $(BUILT_ADD_ON)/ $(DIST)/add-on/
+	mv $(DIST)/add-on/chrome/content/shared/vendor/react-prod.js \
+	   $(DIST)/add-on/chrome/content/shared/vendor/react.js
+
+.PHONY: standalone_dist
+standalone_dist:
+	mkdir -p dist
```
Comment (Mardak) [inline Makefile]: Some `dist` references from moved lines that should be updated to `$(DIST)`

File: Makefile [hunk 5]
```diff
@@ -171,6 +179,53 @@ $(BUILT)/$(ADD-ON)/chrome.manifest: $(ADD-ON)/jar.mn
...
+		< content/index.html > dist/index.html
+
+XPI_FILE := built/$(XPI_NAME)
+
+# so we can type "make xpi" without depend on the file directly
+.PHONY: xpi
+xpi: $(XPI_FILE)
+
+$(XPI_FILE): $(REPO_BIN_DIR)/build_extension.sh build
+	@$(REPO_BIN_DIR)/build_extension.sh $(@F) built add-on
```
Comment (Mardak) [inline Makefile]: Not sure if it's worthwhile to skip the `$(@F)` and just use `$(XPI_NAME)` to match the line below for `dist_xpi` with the only differences between the lines being `$(BUILT)` and `$(DIST)`

**Comment by Standard8** [comment]
@Mardak I've address your review comments and updated the PR. I'll squash the commits before landing once this gets r+.

File: Makefile [hunk 6]
```diff
@@ -120,6 +119,19 @@ vendor_libs: $(BACKBONE_OBJS) $(CLASSNAME_OBJS) $(LODASH_OBJS) $(REACT_OBJS) \
...
+	mkdir -p $(@D)
+	. $(VENV)/bin/activate && \
+	  python $(VENV)/lib/python2.7/site-packages/mozbuild/preprocessor.py \
+					 -D DEBUG=1 -o $@ add-on/preferences/prefs.js
+
+$(DIST)/add-on/chrome/content/preferences/prefs.js: $(VENV) add-on/preferences/prefs.js
+	mkdir -p $(@D)
+	. $(VENV)/bin/activate && \
+	  python $(VENV)/lib/python2.7/site-packages/mozbuild/preprocessor.py \
+					 -o $@ add-on/preferences/prefs.js
```
Comment (Mardak) [inline Makefile]: nit: I believe these want 1 tab and 9 spaces before -D or -o

File: Makefile [hunk 7]
```diff
@@ -171,6 +181,51 @@ $(BUILT)/$(ADD-ON)/chrome.manifest: $(ADD-ON)/jar.mn
...
+	   $(DIST)/add-on/chrome/content/shared/vendor/react.js
+
+.PHONY: standalone_dist
+standalone_dist:
+	mkdir -p $(DIST)
+	$(RSYNC) content/* $(DIST)
+	NODE_ENV="production" $(NODE_LOCAL_BIN)/webpack \
+		-p -v --display-errors
+	sed 's#webappEntryPoint.js#js/standalone.js#' \
+		< content/index.html > $(DIST)/index.html
```
Comment (Mardak) [inline Makefile]: nit: not entirely sure what spacing we want but probably not 2 tabs

**Pull request merged**